File: content/courses/parallel_computing_introduction/solns/mpitrap.py
import sys
import numpy as np
from mpi4py import MPI

# Calculate a definite integral using trapezoid rule
# Does not use scipy so we can parallelize it
                                                                                
def f(x):
    return np.sin(x)

def trap(a,b,h,n,f):
    integral=(f(a) + f(b))/2.0
    x=a
    for i in range(n):
        x+=h
        integral+=f(x)
    integral*=h
    return integral

def main():

    comm=MPI.COMM_WORLD
    nprocs=comm.Get_size()
    rank=comm.Get_rank()

    a=np.empty(1,dtype='float')
    b=np.empty(1,dtype='float')
    n=np.empty(1,dtype='int')
    shutdown=None

    if rank==0:
        if len(sys.argv)==4:
            a[0]=float(sys.argv[1])
            b[0]=float(sys.argv[2])
            n[0]=int(float(sys.argv[3]))
            shutdown=False
        else:
            print("Usage: lower bound, upper bound, number of steps")
            shutdown=True
       #Not going to bother with redistributing on this one
        if (not shutdown):
            if n[0]%nprocs!=0:
                print("The number of processes must evenly divide the number of steps ")
                shutdown=True
    
    #Note lower case and return value
    shutdown=comm.bcast(shutdown,root=0)
    if shutdown:
        MPI.Finalize()
        exit()
    
    comm.Bcast([a,MPI.FLOAT],root=0)
    comm.Bcast([b,MPI.FLOAT],root=0)
    comm.Bcast([n,MPI.INT],root=0)

    h=(b-a)/n
    local_n=n[0]//nprocs
    local_a=a[0]+rank*local_n*h
    local_b=b[0]+rank*local_n*h
    
    integral=np.empty(1,dtype='float')
    integral[0]=trap(local_a,local_b,h,local_n,f)

    total=np.empty(1)
    comm.Reduce(integral,total,op=MPI.SUM)
    
    if rank==0:
        print("Result ",total)
    
    MPI.Finalize()
    
if __name__=="__main__":
    main()

File: content/courses/parallel_computing_introduction/solns/mpitrap.cxx
#include <iostream>
#include <iomanip>
#include <string>
#include <sstream>
#include <cmath>
//For templated function objects
#include <functional>
#include <mpi.h>

using namespace std;

float f(float x) {
  return sin(x);
}

//Function pointer
//float trap(float a, float b, float h, int n, float (*function)(float)) {
//Templated function object
float trap(float a, float b, float h, int n, function<float(float)> f) {

   float integral = (f(a) + f(b))/2.0;
   float x=a;
   for (int i=0; i<n; ++i) {
       x+=h;
       integral += f(x);
   }

   integral *= h;
   return integral;
}

int main(int argc, char **argv) {
                                                                                
// Calculate a definite integral using trapezoid rule
                                                                                
    float  a, b;
    int    n;

    float  h, integral;
    float  x;

    int   nprocs, rank;
    float total;

    MPI_Init(&argc,&argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &nprocs);

    bool shutdown=false;

    if (rank==0) {
        if (argc != 4) {
            cout<<"Usage: lower bound, upper bound, number of steps\n";
	    shutdown=true;
        }
        else {
           stringstream lb;
           lb<<argv[1];
           lb>>a;
           a=stof(argv[1]);

           stringstream ub;
           ub<<argv[1];
           ub>>b;
           b=stof(argv[2]);

           stringstream steps;
           steps<<argv[3];
           steps>>n;
        }
	//Not going to bother to redistribute this one
        if (!shutdown) {
	    if (n%nprocs != 0) {
	        cout<<"Number of processes must evenly divide number of steps\n";
	        shutdown=true;
	    }
	}
    }

    MPI_Bcast(&shutdown,1,MPI_CXX_BOOL,0,MPI_COMM_WORLD);
    if (shutdown) {
        MPI_Finalize();
	exit(1);
    }

    MPI_Bcast(&a,1,MPI_REAL,0,MPI_COMM_WORLD);
    MPI_Bcast(&b,1,MPI_REAL,0,MPI_COMM_WORLD);
    MPI_Bcast(&n,1,MPI_INTEGER,0,MPI_COMM_WORLD);

    h=(b-a)/n;

    int   local_n=n/nprocs;
    float local_a=a+rank*local_n*h;
    float local_b=b+rank*local_n*h;

    integral = trap(local_a,local_b,h,local_n,&f);

    MPI_Reduce(&integral,&total,1,MPI_FLOAT,MPI_SUM,0,MPI_COMM_WORLD);

    if (rank==0) {
        cout<<" Result "<<total<<endl;
    }

}

